### PR TITLE
Create Workspaces Optimistically Client-side

### PR DESCRIPTION
**Pull request by thienlnam**
### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/224690

### Tests

Online flow:

1. Create a workspace while online
2. Notice that the default chats are created and you are navigate to the workspace settings
3. Check that you can adjust the workspace settings and everything works as expected

Offline flow:
1. Create a workspace while offline
2. Notice that the default chats are also created and you are still navigated to the workspace settings
3. Adjust the workspace name while offline and notice that it is updated successfully
4. Go back online and ensure that everything syncs with the server

- [ ] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps

Online flow:

1. Create a workspace while online
2. Notice that the default chats are created and you are navigate to the workspace settings
3. Check that you can adjust the workspace settings and everything works as expected

Offline flow:
1. Create a workspace while offline
2. Notice that the default chats are also created and you are still navigated to the workspace settings
3. Adjust the workspace name while offline and notice that it is updated successfully
4. Go back online and ensure that everything syncs with the server
- [ ] Verify that no errors appear in the JS console

### CODE REVIEW

File: src/libs/actions/Policy.js [hunk 1]
```diff
@@ -802,6 +737,110 @@ function generatePolicyID() {
     return _.times(16, () => Math.floor(Math.random() * 16).toString(16)).join('').toUpperCase();
 }
 
+/**
+ * Optimistically creates a new workspace and default workspace chats
+ * @param {*} ownerEmail
```
Comment (marcochavezf) [inline src/libs/actions/Policy.js:742]: There's no param `ownerEmail` for `createWorkspace`
```suggestion
```

File: src/libs/actions/Report.js [hunk 2]
```diff
@@ -649,6 +649,46 @@ function createOptimisticChatReport(participantList) {
     };
 }
 
+function createOptimisticWorkspaceChats(policyID, ownerEmail) {
```
Comment (marcochavezf) [inline src/libs/actions/Report.js:652]:
```suggestion
/**
 * @param {String} policyID
 * @param {String} ownerEmail
 * @returns {*}
 */
function createOptimisticWorkspaceChats(policyID, ownerEmail) {
```

**Review by marcochavezf — CHANGES_REQUESTED** [review]
NAR: looks good to me! just a few comments related to jsdoc

**Comment by PauloGasparSv** [comment]
Are you guys able to create the workspace offline, then go online and open the created workspace chats? Not sure if this is in the scope of the tests or if it's only happening here but: 
The workspace is being created and updated successfully but the chats are not (only one of them). But the chats to reflect any changes I do in the workspace name (as you can see, mine are saying potato)
image

File: src/libs/actions/Policy.js [hunk 3]
```diff
@@ -802,6 +737,109 @@ function generatePolicyID() {
...
+                role: CONST.POLICY.ROLE.ADMIN,
+                owner: sessionEmail,
+                outputCurrency: 'USD',
+                pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.ADD,
+            },
+        },
+        {
+            onyxMethod: CONST.ONYX.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY_MEMBER_LIST}${policyID}`,
+            value: getSimplifiedEmployeeList([{email: sessionEmail, role: CONST.POLICY.ROLE.ADMIN}]),
```
Comment (mountiny) [inline src/libs/actions/Policy.js:778]: Theoretically we should not need this method in App going forward as we used to to massage data coming from server and that should already be coming in the correct shape.

Should we just already se this object to the correct format and avoid using this data format manipulating method?

**Review by mountiny — COMMENTED** [review]
Great job on this one, havent tested this locally but left one comment

File: src/libs/actions/Policy.js (hunk 3 above)
Comment (thienlnam) [reply src/libs/actions/Policy.js:778]: Do you have an issue link / PR for this? 

I am thinking we still need it locally if they want to update their workspace members offline, but then once we get online the changes should sync with the server and update the list

**Comment by thienlnam** [comment]
> The workspace is being created and updated successfully but the chats are not (only one of them). But the chats to reflect any changes I do in the workspace name (as you can see, mine are saying potato)

I do notice the skeleton UI is still there sometimes, going to dig a bit and see if I can stop this from happening

> Followup: Also is it duplicating the chats sometimes?

This I haven't run into yet

File: src/libs/actions/Policy.js (hunk 3 above)
Comment (mountiny) [reply src/libs/actions/Policy.js:778]: I dont have an issue for this but I am just coming from the original refactoring ides to do no data manipulation on the frontend. That said, it concerns data getting from backend so in this case it is fine, but I am just wondering if we could already format this into the shape we need (given it is just one email and role it should be fine) to simplify stuff.

I am not suggesting to deprecate the method and delete it from the codebase, just not use it here :D

Comment (thienlnam) [reply src/libs/actions/Policy.js:778]: Ah gotcha, yeah I can do that 👍

Comment (mountiny) [reply src/libs/actions/Policy.js:778]: Nice thanks!

**Comment by thienlnam** [comment]
This is ready for another set of reviews!

cc @PauloGasparSv @mountiny

**Comment by thienlnam** [comment]
> There is something wrong with the checklist test.

I think that's because there isn't a C+ that is reviewing this PR and completing the checklist 😆

File: src/libs/actions/Report.js [hunk 4]
```diff
@@ -616,6 +616,197 @@ function createOptimisticChatReport(participantList) {
...
+        notificationPreference: '',
+        oldPolicyName: policyName,
+        ownerEmail: '__FAKE__',
+        participants: [currentUserEmail],
+        stateNum: 0,
+        statusNum: 0,
+        visibility: undefined,
+    };
+    const announceReportActionData = {
+        0: {
```
Comment (PauloGasparSv) [inline src/libs/actions/Report.js:652]: Is there a reason we are not using an array? Don't think it's a problem, just curious : )

File: src/libs/actions/Report.js [hunk 5]
```diff
@@ -616,6 +616,197 @@ function createOptimisticChatReport(participantList) {
...
+        oldPolicyName: policyName,
+        ownerEmail: '__FAKE__',
+        participants: [currentUserEmail],
+        stateNum: 0,
+        statusNum: 0,
+        visibility: undefined,
+    };
+
+    const adminsReportActionData = {
+        0: {
```
Comment (PauloGasparSv) [inline src/libs/actions/Report.js:710]: > Is there a reason we are not using an array? Don't think it's a problem, just curious : )

Here too

File: src/libs/actions/Report.js (hunk 4 above)
Comment (thienlnam) [reply src/libs/actions/Report.js:652]: Onyx is built to work with objects not arrays - we've been trying to update any current array usage to objects

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Looking good. After I test it out I will leave another review if I notice anything weird. Did you add the pattern B offline feedback? Please add to the test / QA steps a section for the transition flows that have to do with creating workspaces, which can be found [here](https://expensify.testrail.io/index.php?/suites/view/518&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=228827).

**Review by neil-marcellini — COMMENTED** [review]
It looks like the browser notification is breaking due to it missing a title from the report action person [here](https://github.com/Expensify/App/blob/68c20364e5bef3fbaf3ce885b4a1d7418ea93e63/src/libs/Notification/LocalNotification/BrowserNotifications.js#L115). There's also an error about Onyx.merge failing.

I'm also seeing the not found page show up briefly.
Here's a video:

https://user-images.githubusercontent.com/26260477/188491222-2741bbcc-1510-4748-8641-1dc4f8ec703c.mov

File: src/libs/actions/Policy.js [hunk 6]
```diff
@@ -802,6 +737,136 @@ function generatePolicyID() {
...
+                role: CONST.POLICY.ROLE.ADMIN,
+                owner: sessionEmail,
+                outputCurrency: 'USD',
+                pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.ADD,
+            },
+        },
+        {
+            onyxMethod: CONST.ONYX.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY_MEMBER_LIST}${policyID}`,
+            value: {[sessionEmail]: {}},
```
[the remark this answers is not in the document]
Comment (thienlnam) [reply src/libs/actions/Policy.js]: Added some items

**Comment by thienlnam** [comment]
I've fixed a lot of bugs that were found, this is ready to be review again

cc @srikarparsi @stitesExpensify @marcochavezf @mountiny @neil-marcellini @PauloGasparSv

**Comment by thienlnam** [comment]
This can be reviewed / corresponding web PR is deploying to prod today so we can merge this

**Review by mountiny — DISMISSED** [review]
Code looks good to me, great job on this one! Havent got around to test right now, though

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Thanks for updating, the code is looking a lot cleaner now. When testing on web I'm still having problems with the "Not found" page showing briefly. It's strange that it happens even with `makeRequestWithSideEffects`. Do you also see this problem when testing?

File: src/libs/actions/Policy.js [hunk 7]
```diff
@@ -754,14 +755,19 @@ function createWorkspace() {
         expenseChatData,
         expenseReportActionData,
     } = Report.createOptimisticWorkspaceChats(policyID, workspaceName);
-    API.write('CreateWorkspace', {
+
+    // We need to use makeRequestWithSideEffects as we try to redirect to the policy right after creation
+    // The policy hasn't been merged in Onyx data at this point, leading to an intermittent Not Found screen
+    // eslint-disable-next-line rulesdir/no-api-side-effects-method
+    API.makeRequestWithSideEffects('CreateWorkspace', {
```
[the remark this answers is not in the document]
Comment (neil-marcellini) [reply src/libs/actions/Policy.js:762]: After testing it looks like this isn't fixing the problem. Weird.

**Comment by neil-marcellini** [comment]
There's also merge conflicts now 😕

File: src/libs/actions/Policy.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (neil-marcellini) [reply src/libs/actions/Policy.js:762]: Here's a video for proof
https://user-images.githubusercontent.com/26260477/188608717-46073ad2-4374-4d71-b978-224cf30b4c99.mov

**Comment by thienlnam** [comment]
> When testing on web I'm still having problems with the "Not found" page showing briefly. It's strange that it happens even with makeRequestWithSideEffects. Do you also see this problem when testing?

Hmm, I had this problem before but after adding makeRequestWithSideEffects this has not shown up for me.  It seems to just be because we are trying to navigate too quickly and it's it might be that we're trying to merge too many items

File: src/pages/workspace/withFullPolicy.js [hunk 8]
```diff
@@ -59,9 +59,6 @@ const fullPolicyPropTypes = {
         /** The URL for the policy avatar */
         avatar: PropTypes.string,
 
-        /** A list of emails for the employees on the policy */
-        employeeList: PropTypes.arrayOf(PropTypes.string),
```
[the remark this answers is not in the document]
Comment (thienlnam) [reply src/pages/workspace/withFullPolicy.js:63]: Now we have a seperate onyx key for employeeList, so this isn't needed anymore - it was throwing an error

File: src/libs/actions/Policy.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (thienlnam) [reply src/libs/actions/Policy.js:762]: Are you able to try after clearing onyx data? From the console log it seems like there are other problems you are running into - I haven't seen that when testing on web locally

[the remark this answers is not in the document]
Comment (neil-marcellini) [reply src/libs/actions/Policy.js:762]: Ah good call. I'll try that.

[the remark this answers is not in the document]
Comment (thienlnam) [reply src/libs/actions/Policy.js:762]: Resolving as this seems to be an issue with @neil-marcellini 's dev environment

**Review by neil-marcellini — APPROVED** [review]
The errors I am seeing are because my dev environment is having unrelated problems. Others have confirmed that it's working well. Good to go!

**Review by mountiny — APPROVED** [review]

File: src/libs/actions/Report.js [hunk 9]
```diff
@@ -588,13 +588,27 @@ function fetchAllReports(
...
+ * @param {String} oldPolicyName
  * @returns {Object}
  */
-function createOptimisticChatReport(participantList) {
+function createOptimisticChatReport(
+    participantList,
+    reportName = 'Chat Report',
+    chatType = '',
+    policyID = '_FAKE_',
+    ownerEmail = '__FAKE__',
```
Comment (stitesExpensify) [inline src/libs/actions/Report.js:604]: is this just an existing thing that we do? policyID has one underscore and this has 2

Comment (stitesExpensify) [reply src/libs/actions/Report.js:604]: oh it must be because i wrote this originally and copy/pasted the response :lolsob:

**Review by stitesExpensify — APPROVED** [review]

**Review by marcochavezf — APPROVED** [review]

**Comment by thienlnam** [comment]
Everything passed

**Comment by MonilBhavsar** [comment]
I'm coming from this issue https://github.com/Expensify/App/issues/12151 and as a part of checklist item, commenting here

> The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

Looks like we missed adding red brick road for workspace chat rooms created offline.

File: src/libs/actions/Policy.js [hunk 10]
```diff
@@ -802,6 +737,155 @@ function generatePolicyID() {
...
+        optimisticData: [{
+            onyxMethod: CONST.ONYX.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
+            value: {
+                id: policyID,
+                type: CONST.POLICY.TYPE.FREE,
+                name: workspaceName,
+                role: CONST.POLICY.ROLE.ADMIN,
+                owner: sessionEmail,
+                outputCurrency: 'USD',
```
Comment (mollfpr) [inline src/libs/actions/Policy.js:780]: ✋ Coming from https://github.com/Expensify/App/issues/12047

We should set the currency according to the current account currency.

**Pull request merged**